packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
-
-      for (const action of sortedActions) {
-        // start with inputs
-        const inputArtifacts = action.inputs;
-        for (const inputArtifact of inputArtifacts) {
-          if (!inputArtifact.artifactName) {
-            ret.push(`Action '${action.actionName}' has an unnamed input Artifact that's not used as an output`);
-          } else if (!outputArtifactNames.has(inputArtifact.artifactName)) {
-            ret.push(`Artifact '${inputArtifact.artifactName}' was used as input before being used as output`);
+    interface PipelineLocation {
+      stageIndex: number;
+      stageName: string;
+      action: FullActionDescriptor;
+    }
+
+    /**
+     * Returns whether a is before or the same order as b
+     */
+    function beforeOrEqual(a: PipelineLocation, b: PipelineLocation) {
+      if (a.stageIndex !== b.stageIndex) { return a.stageIndex < b.stageIndex; }
+      return a.action.runOrder <= b.action.runOrder;
+    }
+
+    function renderLoc(loc: PipelineLocation) {
+      // runOrders are 1-based, so make the stageIndex also 1-based otherwise it's going to be confusing.
+      return `stage ${loc.stageIndex + 1} ('${loc.stageName}') action ${loc.action.runOrder} ('${loc.action.actionName}')`;
+    }
+
+    const producer: Record<string, PipelineLocation> = {};
+    const firstConsumer: Record<string, PipelineLocation> = {};
+
+    for (const [stageIndex, stage] of enumerate(this._stages)) {
+      // For every output artifact, get the producer
+      for (const action of stage.actionDescriptors) {
+        const thisLocation = { stageIndex, action, stageName: stage.stageName };
+
+        for (const outputArtifact of action.outputs) {
+          // output Artifacts always have a name set
+          const name = outputArtifact.artifactName!;
+          if (producer[name]) {
+            ret.push(`Artifact '${name}' is used as an output by both '${producer[name].action.actionName}' and '${action.actionName}'. Every artifact can only be produced once.`);
+            continue;
           }
+
+          producer[name] = thisLocation;
         }
 
-        // then process outputs by adding them to the Set
-        const outputArtifacts = action.outputs;
-        for (const outputArtifact of outputArtifacts) {
-          // output Artifacts always have a name set
-          if (outputArtifactNames.has(outputArtifact.artifactName!)) {
-            ret.push(`Artifact '${outputArtifact.artifactName}' has been used as an output more than once`);
-          } else {
-            outputArtifactNames.add(outputArtifact.artifactName!);
+        // For every input artifact, get the first consumer
+        for (const inputArtifact of action.inputs) {
+          const name = inputArtifact.artifactName;
+          if (!name) {
+            ret.push(`Action '${action.actionName}' has an unnamed input Artifact (probably not used as an output in this pipeline)`);
```

Comment:
Why change this message here - why the "probably"? There's nothing probable about that, it's a certainty 🙂. I would say leave it what it was before: `Action '${action.actionName}' has an unnamed input Artifact that's not used as an output`.